firehose.go

```diff
@@ -0,0 +1,93 @@
+package hedwig
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/binary"
+	"io"
+)
+
+type Firehose struct {
+	messageValidator *messageValidator
+}
+
+func (f *Firehose) Deserialize(reader io.Reader) ([]Message, error) {
+	// TODO: loop through contents and return messages
+	var messagePayload []byte
+	runWithTransportMessageAttributes := false
+	var messages []Message
+	if f.messageValidator.encoder.IsBinary() {
+		for {
+			// TLV format: 8 bytes for size of message, n bytes for the actual message
+			msgSize := make([]byte, 8)
+			_, err := reader.Read(msgSize)
+			if err == io.EOF {
+				break
+			} else if err != nil {
+				return nil, err
+			}
+			msgLength := binary.LittleEndian.Uint64(msgSize)
+			messagePayload = make([]byte, msgLength)
+			_, err = reader.Read(messagePayload)
+			if err != nil && err != io.EOF {
+				return nil, err
+			}
+
+			res, err := f.messageValidator.deserialize(messagePayload, nil, nil, &runWithTransportMessageAttributes)
+			if err != nil {
+				return nil, err
+			}
+			messages = append(messages, *res)
+		}
+	} else {
+		bf := bufio.NewReader(reader)
+		for {
+			c, err := bf.ReadBytes('\n')
```

Comment:
looking at bufio implementation it looks like this allocates a new buffer for every call :( It's possible that scanner is more appropriate here, but lets put a todo and tackle that as a separate PR